### DISCUSSION TITLE
[WIP] changes for next release 3.11.2

### DISCUSSION THREAD

**Discussion by nlohmann** [nlohmann/json#3649] [Announcements]
This thread collects all changes made since the 3.11.1 (#3647) release.

#### Bug fixes

- Fix the [`value`](https://json.nlohmann.me/api/basic_json/value/) function which was broken for strings, size types, and `nullptr` in release 3.11.0. #3652 #3655 #3663
- Fix the `json_fwd.hpp` header to be self-contained and add it to the single-header release. #3656 #3679 #3687
- Fix regression that broke using [`json_pointer`](https://json.nlohmann.me/api/json_pointer/) as key in associative containers. #3680 #3685
- Add missing constraint to deprecated JSON Pointer overloads of [`contains`](https://json.nlohmann.me/api/basic_json/contains/) and [`at`](https://json.nlohmann.me/api/basic_json/at/). #3658 #3681 
- Fix comparison between [`json_pointer`](https://json.nlohmann.me/api/json_pointer/) and strings with `==` and `!=`. These comparisons worked in 3.10.5, but were broken in 3.11.0 and 3.11.1. #3654 #3664
- Fix `to_json` conversion of `std::vector<bool>::reference` and `std::vector<bool>::const_reference` for STLs where these are the same as `basic_json::boolean_t&` and `basic_json::boolean_t`, respectively. #3677 #3678

#### Improvements

- Restructure inline namespace and allow version component to be disabled. See [documentation](https://json.nlohmann.me/features/namespace/). #3683 #3696 #3697 #3698
- Avoid heap allocations in [BJData](https://json.nlohmann.me/features/binary_formats/bjdata/) parser. #3637

#### Documentation

- Publish documentation on every push to `develop` branch. #3660 #3673
- Add missing examples for the [public API](https://json.nlohmann.me/api/basic_json/). #3672 #3686
- Fix typo in [`json_pointer`](https://json.nlohmann.me/api/json_pointer/) documentation. #3692

#### Community

- Add a badge for the [Discord chat](https://discord.gg/6mrGXKvX7y) to the README file. The goal of that additional communication channel beyond [the existing ones](https://github.com/nlohmann/json#support) is to quickly coordinate between the contributors. #3651
- Complete contributor list. #3662 #3670

#### CI

- Remove the `macos-10.15` image from the CI as it is removed by GitHub Actions. #3612 #3615 #3626
- Remove hardcoded paths in Ubuntu workflow. #3626
- Only trigger AppVeyor builds if relevant files have been changed. #3626
- Fix CodeQL warning. #3626
- Harmonize naming of GitHub Actions jobs. #3661
- Add labeler action to automatically add PR labels based on the changed files. #3671 #3674 #3675
- Lint API documentation in the CI. #3672
- Add local [LGTM](http://lgtm.com) configuration and suppress warnings on third-party scripts. #3643

#### Deprecations

- This release deprecations comparing JSON Pointers with strings via [`operator==`](https://json.nlohmann.me/api/json_pointer/operator_eq/) and [`operator!=`](https://json.nlohmann.me/api/json_pointer/operator_ne/). To compare a [`json_pointer`](https://json.nlohmann.me/api/json_pointer/) `p` with a string `s`, convert `s` to a `json_pointer` first and use [`json_pointer::operator==`](https://json.nlohmann.me/api/json_pointer/operator_eq/) or [`json_pointer::operator!=`](https://json.nlohmann.me/api/json_pointer/operator_ne/). #3684

**Comment by falbrechtskirchinger** [comment]
Here's my list of PRs that are ready to go (possibly pending CI run), starting with the easy ones:
* #3681
* #3678
* #3679
* #3643

Easy but may need discussion:
* #3684
* #3685

These are the bigger/more complicated ones:
* #3663
* #3683

**Comment by falbrechtskirchinger** [comment]
**Bug fixes**, first sentence 3.10.0 -> 3.11.0:
> Fix the value function which was broken for string and size types in release 3.1**1**.0

Last sentence:
Fix to_json conversion of \``std::vector<bool>::reference`\` and \``std::vector<bool>::const_reference`\` for STLs where these are the same as \``basic_json::boolean_t&`\` and \``basic_json::boolean_t`\`, respectively.

(With extra backticks for your convenience :-))

**Reply by nlohmann** [reply]
I appreciate the extra backticks :)

**Comment by steffenb7333** [comment]
Please do not forget to update nlohmann_jsonConfigVersion.cmake.in (https://github.com/nlohmann/json/issues/3657#issuecomment-1203758621).

**Reply by falbrechtskirchinger** [reply]
I've thought about this some more. We do follow same-major-version compatibility in the sense that existing code continues to compile with newer versions of the library. Therefore I believe `nlohmann_jsonConfigVersion.cmake.in` is still correct (and changing it would also break more than it would fix).

**Reply by steffenb7333** [reply]
I see that there is room for a discussion about what makes a two versions of a header-only library compatible in relation to another, but I have some reservations about having a version-encoded-namespace by default on the one hand and calling that having same-major-version-compatibility. Because on the other hand as soon as I expose anything from the `nlohmann`-namespace in the ABI of a library using `json` I am back to exact-version-semantics (ignoring the namespace-hack).

And just using any newest version on the sourcecode level can be easy achieved by
```
set(CMAKE_FIND_PACKAGE_SORT_ORDER NATURAL)
set(CMAKE_FIND_PACKAGE_SORT_DIRECTION DEC)
find_package(nlohmann_json)
```

Addendum:
The compatibility-check is mostly only relevant for [find_dependency](https://cmake.org/cmake/help/latest/module/CMakeFindDependencyMacro.html), therefore the compatibility when linking code is relevant.

**Reply by steffenb7333** [reply]
Just to be clear, the problem I see with the current state is that it is inconsistent:

- As an application-developer I am getting major-version-compatibility
- As a header-only library developer I am also getting major-version-compatibility
- As a library developer creating a non-header-only-library and exposing anything from the library in the ABI - even by accident - I am getting same-version-compatibility
- As a user of one of the libraries described above I inherit - maybe by accident - whatever what was valid for the library or same-version-compatibility depending on circumstances
- As a maintainer for multiple packages depending on each other I cannot even express that in a sane way without recompiling everything and anything using `json`.

**Reply by falbrechtskirchinger** [reply]
And because of those inconsistencies, I feel same-version-compatibility would be unnecessarily limiting. :man_shrugging:

> * As a maintainer for multiple packages depending on each other I cannot even express that in a sane way without recompiling everything and anything using `json`.

Yup.

**Reply by steffenb7333** [reply]
A change of `nlohmann_jsonConfigVersion.cmake.in` towards same-version-compatibility would basically only affect packages citing `json` as a dependency via `find_dependency` and remove the inconsistency explained. The "normal" user, using the package via `find_package` would not be affected by the change.
Basically the current documented state is (API=major-version-compatibility, ABI=major-version-compatibility) whereas the reality is (API=major-version-compatibility, ABI=same-version-compatibility).

**Reply by falbrechtskirchinger** [reply]
```cmake
find_package(nlohmann_json 3.9.0 REQUIRED)
```

I request version 3.9.0 because it introduced a feature I want to use. Newer versions would work for me as well. If v3.10.5 is installed, this line would fail with same-version-compatibility, would it not?

**Reply by steffenb7333** [reply]
You are correct and I see your point.
Nevertheless the current situation either breaks compiled libraries with `nlohmann`-symbols (and a later library update) or versioned `find_package`-calls.

I am out of ideas ATM.

**Reply by falbrechtskirchinger** [reply]
Indeed. Just note that things were already broken and the inline namespace will prevent known segfault scenarios.

**Comment by falbrechtskirchinger** [comment]
> Fix regression that broke using json_pointer as object keys.

"as map key" or "as key in associative containers"
